### PR TITLE
3208 share from catalog

### PR DESCRIPTION
**Pull request by soyarsauce**
~draft for ci to resolve 3208~
Ready for review

### CODE REVIEW

**Comment by soyarsauce** [comment]
See comment @ https://github.com/TerriaJS/terriajs/issues/3208#issuecomment-485657047 for the todo overhaul comments

File: lib/ReactViews/ExplorerWindow/ExplorerWindow.jsx
Comment (chloeleichen) [inline lib/ReactViews/ExplorerWindow/ExplorerWindow.jsx]: we don't actually need this coz the `viewCatalogMember` function in `viewState.js` actually check if the `previewed`  item is added by user or from the default data explorer

File: lib/ReactViews/ExplorerWindow/ExplorerWindow.jsx
Comment (chloeleichen) [inline lib/ReactViews/ExplorerWindow/ExplorerWindow.jsx]: we have a problem here. If share panel on the top right hand corner is open and explorer window is open, esc key close the share panel, and further esc key press does nothing. I think this behaviour is problematic. Ideally the order should be: close catalog share -> close explorer window -> close map share panel

File: lib/ReactViews/Map/Panels/SharePanel/BuildShareLink.js
Comment (chloeleichen) [inline lib/ReactViews/Map/Panels/SharePanel/BuildShareLink.js]: Not super convinced that we need to introduce `viewState` into BuildShareLink, though i guess non of these will be problem once we are in mobx land

**Review by chloeleichen — CHANGES_REQUESTED** [review]
Thanks @soyarsauce

File: lib/Models/Terria.js
Comment (soyarsauce) [reply lib/Models/Terria.js]: 👍

File: lib/Models/Terria.js
Comment (soyarsauce) [reply lib/Models/Terria.js]: 👍

File: lib/ReactViews/ExplorerWindow/ExplorerWindow.jsx
Comment (soyarsauce) [reply lib/ReactViews/ExplorerWindow/ExplorerWindow.jsx]: Thanks! This also actually prompted me to check and catalog-shareable-user-added-shares were indeed not working, so I've fixed that w/ https://github.com/TerriaJS/terriajs/pull/3278/commits/19128b5edec86e58e35c86c5d741951872c1e6d0 too

File: lib/ReactViews/ExplorerWindow/ExplorerWindow.jsx
Comment (soyarsauce) [reply lib/ReactViews/ExplorerWindow/ExplorerWindow.jsx]: Thanks for picking up on this too, the top-right-share-panel open + explorer window combination is fixed now but it also gets dismissed upon any further interaction inside the explorer window.

The end result is esc keys will always eventually bring you back to the map without modals open, but because we don't have any modal layer logic yet are you happy to leave it like this, then we raise an issue to create the new modal system post-mobx?

File: lib/ReactViews/Map/Panels/SharePanel/BuildShareLink.js
Comment (soyarsauce) [reply lib/ReactViews/Map/Panels/SharePanel/BuildShareLink.js]: We need some way to tell if the explorer window is open via `explorerPanelIsVisible` so I can't think of any other way since there's no reference from `viewState` -> `terria`? So yes hopefully mobx land simplifies this

File: lib/ReactViews/Map/Panels/SharePanel/share-panel.scss
Comment (soyarsauce) [reply lib/ReactViews/Map/Panels/SharePanel/share-panel.scss:60]: I've reverted the original share panel back to a dark input & fixed up the contrast for that embed-text-color w/ Phil's OK as well.

File: lib/ReactViews/Map/Panels/SharePanel/SharePanel.jsx
Comment (kring) [inline lib/ReactViews/Map/Panels/SharePanel/SharePanel.jsx]: Why "for print"?

File: lib/ReactViews/Preview/GroupPreview.jsx
Comment (kring) [inline lib/ReactViews/Preview/GroupPreview.jsx]: Fancy!

File: lib/ReactViews/Preview/measureWidth.js
Comment (kring) [inline lib/ReactViews/Preview/measureWidth.js]: Sounds like this originally came from another project. Are we definitely ok to use it?

File: test/ReactViews/Map/Panels/SharePanel/BuildShareLinkSpec.js
Comment (kring) [inline test/ReactViews/Map/Panels/SharePanel/BuildShareLinkSpec.js]: I think you may need to use `done.fail` instead of just `fail` here, otherwise you'll get a timeout rather than a quick failure if the promise rejects.

Thanks for writing tests for this by the way! :)

**Review by kring — CHANGES_REQUESTED** [review]
Looking good! Just a few things to fix up.

File: lib/ReactViews/Map/Panels/SharePanel/BuildShareLink.js
Comment (soyarsauce) [reply lib/ReactViews/Map/Panels/SharePanel/BuildShareLink.js]: Thanks - have re-written the logic to work with this case as well now, & another test to really makes sure in these conditions that `previewedItemId` doesn't get added to the share url @ https://github.com/TerriaJS/terriajs/pull/3278/commits/36cae5343a4eb1558e4db2914c3f88530d34e2c5#diff-ad2a0224e6b8153d6618d26e64308cf9R59

File: lib/Models/Terria.js
Comment (soyarsauce) [reply lib/Models/Terria.js]: Thanks - it was used, then not, and now finally used again, almost not required but in the earlier mentioned case it will come in handy

File: lib/Models/Terria.js
Comment (soyarsauce) [reply lib/Models/Terria.js]: This is great - moved everything over to the `ViewState` & removed all changes to ExplorerWindow @ https://github.com/TerriaJS/terriajs/pull/3278/commits/9db8f6159881d2d3d97251a8e0089384b17f777d

File: lib/ReactViews/Map/Panels/SharePanel/SharePanel.jsx
Comment (soyarsauce) [reply lib/ReactViews/Map/Panels/SharePanel/SharePanel.jsx]: SharePanel still gets used with the print/embed options via the top right icons on the map, "for print" probably wasn't the best name 😃 https://github.com/TerriaJS/terriajs/pull/3278/commits/45d2c3697a184b16f07d9ac88c0627b8f3b67742

File: lib/ReactViews/Preview/measureWidth.js
Comment (soyarsauce) [reply lib/ReactViews/Preview/measureWidth.js]: I originally wrote it for another project and tied it in quite heavily to redux - so it's been adapted to what we have here, but I will confirm the definitely part & report back - in the meantime I may just make it work with both styles of refs as it's not like the callback ref has been deprecated.

File: test/ReactViews/Map/Panels/SharePanel/BuildShareLinkSpec.js
Comment (soyarsauce) [reply test/ReactViews/Map/Panels/SharePanel/BuildShareLinkSpec.js]: Thanks for the pickup 👍

**Comment by soyarsauce** [comment]
Addressed everything now but just waiting for the HoC-from-another-project confirmation

File: lib/ReactViews/Preview/measureWidth.js
Comment (kring) [reply lib/ReactViews/Preview/measureWidth.js]: I'm not too fussed about the callback versus createRef, just wanted to make sure there are no IP issues.

Comment (soyarsauce) [reply lib/ReactViews/Preview/measureWidth.js]: Yes hence the confirming the definitely part :) Should get an answer tomorrow.

Comment (soyarsauce) [reply lib/ReactViews/Preview/measureWidth.js]: "of course!"
No issues here!

File: lib/ReactViewModels/ViewState.js
Comment (kring) [inline lib/ReactViewModels/ViewState.js]: It may be equivalent in this case, but this code is unusual. I'd think you would either want `if (defined(previewedItemId) && defined(terria.catalog.shareKeyIndex[previewedItemId]))` or `if (previewedItemId && terria.catalog.shareKeyIndex[previewedItemId])`. Using `defined` on the result of a logical `and` is not typically done.

File: lib/ReactViewModels/ViewState.js
Comment (kring) [inline lib/ReactViewModels/ViewState.js]: We don't care if it's true? Only that it's defined?

File: lib/ReactViewModels/ViewState.js
Comment (soyarsauce) [reply lib/ReactViewModels/ViewState.js]: Oops this certainly wasn't intended & was going for your latter, I just left off a closing paren.

File: lib/ReactViewModels/ViewState.js
Comment (soyarsauce) [reply lib/ReactViewModels/ViewState.js]: Thanks good pickup, I think my idea was the property only ever gets set if it is shared, but I've also now updated to check it's also true

**Comment by soyarsauce** [comment]
I need to do some changes for mobile.

**Comment by soyarsauce** [comment]
@philipgrimmett is happy for it to be disabled until he thinks of a mobile solution, so it has been disabled via `useSmallScreenInterface`.

**Comment by kring** [comment]
This looks good to me. Do you want to review again before we merge @chloeleichen?

**Review by kring — APPROVED** [review]

**Review by chloeleichen — APPROVED** [review]
It works/looks awesome!

**Pull request merged**
